### ISSUE TITLE
Reproducing the results of segmentation

### ISSUE THREAD

**Issue by LYY-HH** [xulianuwa/MCTformer#26]
I am unable to reproduce the segmentation retraining results in your paper (71.9). I have used the pseudo-mask and pre-training weights you provided. I have also trained with exact reference to the default parameters you set in train_seg.py as well as run_seg.sh. But I only get 71.1 results. I am confused about this, can you tell me why? Thanks!

**Comment by Gptao** [comment]
Yes, the code's reproduction is not as description in the paper. Under the same environment and with the provided pseudo-mask, I get 71.1 mIoU of PascalVOC val set too.  Further, I don't understand the superiority of provided pseudo-mask, for the evaluation is not better with mIoU and other metrics.

**Comment by Gptao** [comment]
I suggest that the author should, in a responsible manner, explain the advantages of provided pseudo labels and give a reproduction of the performance provided by the paper  from seed to semseg results under the fixed seed(as current settings).

**Comment by Gptao** [comment]
<img width="579" alt="image" src="https://user-images.githubusercontent.com/32570671/203507857-96f6477a-d5a6-42d9-a916-32e1c003a8be.png">
The above is the evaluation of the paper. The parentheses are the results obtained by training the segmentation model with the provided pseudo labels and the provided deeplabv1 code. The following is the results obtained from the start.

**Comment by LYY-HH** [comment]
I have tried to set the seed in the train_seg.py like the main.py in your repo setup and I get the segmentation result of 68.4% without post-processing, but after adding the post-processing the final result is only 71.6%.

**Comment by xulianuwa** [comment]
> I have tried to set the seed in the train_seg.py like the main.py in your repo setup and I get the segmentation result of 68.4% without post-processing, but after adding the post-processing the final result is only 71.6%.

I have tried multiple ways of setting random seeds (including the one used in main.py) for training the segmentation model, but the loss still changes at different runs. The trained models of different runs can reach comparable segmentation performances without post-processing, but they could have different per class IoU results, for which the same post-processing method (multi-scale testing + CRF) does not always bring same gains. Btw, I'm now working on the problem with the random seed and will update the code once I fix it. Besides, I suggest when you compare your method with ours, you can compare your segmentation results with ours under the same condition of not using post-processing.

**Comment by xulianuwa** [comment]
> Is the default command : python seg/infer_seg.py --weights --network resnet38_seg --list_path voc12/val_id.txt --gt_path --img_path --save_path val_ms_crf --save_path_c val_ms_crf_c --scales 0.5 0.75 1.0 1.25 1.5 --use_crf True contain the post-processing? Otherwise, how to add the post-processing method?

Yes, it used post-processing. Can you change the scales to 1.0 and remove --use_crf to see the results without post-processing?

**Comment by Gptao** [comment]
My current understanding and doubts are:

(1) Under the pseudo tag provided by you, when the seed is fixed to 0, there may be a better result 71.6 than the result 71.1 reproduced by me, as in LYY-HH, so the seed should be fixed in the segmentation code.

(2) Do you fix the seed as 0 in the whole classifier, psa and seg training during code implementation? Why is this part of code removed when open source?

(3) I try to find that performance reproduction other than using pseudo label training segmentation, such as pseudo label mIoU, also fails to achieve the performance of the paper. Is this caused by the random seeds of other parts of the code, such as PSA?

(4) Why are performance so sensitive to random seeds?

I'd appreciate it if you could share your views!

**Comment by xulianuwa** [comment]
> My current understanding and doubts are:
> 
> (1) Under the pseudo tag provided by you, when the seed is fixed to 0, there may be a better result 71.6 than the result 71.1 reproduced by me, as in LYY-HH, so the seed should be fixed in the segmentation code.
> 
> (2) Do you fix the seed as 0 in the whole classifier, psa and seg training during code implementation? Why is this part of code removed when open source?
> 
> (3) I try to find that performance reproduction other than using pseudo label training segmentation, such as pseudo label mIoU, also fails to achieve the performance of the paper. Is this caused by the random seeds of other parts of the code, such as PSA?
> 
> (4) Why are performance so sensitive to random seeds?
> 
> I'd appreciate it if you could share your views!

The main point is not about random seed. Setting or not setting the random seed, the segmentation model can reach around 68.2%. It is the post-processing that cannot give a stable gain, which is generally within a range between 2-4%.

**Comment by Gptao** [comment]
Thank you for your reply,  but I still want to know "It is the post-processing that cannot give a stable gain, which is generally within a range between 2-4%"  is the post-processing’s gain float  for the same trained model (weight)  or  for different model with same performance without post-processing?

**Comment by Gptao** [comment]
The performance gain for the same trained model (weight) is the same. As far as I know, the sem seg performance gain of post-processing including MS+CRF for other method(IRN,AMN,AFA)is far less than 2-4%, and your 2-4% improvement seems not a common result and is hard to obtain for pseudo labels generated with provided code.

**Comment by xulianuwa** [comment]
> The performance gain for the same trained model (weight) is the same. As far as I know, the sem seg performance gain of post-processing including MS+CRF for other method(IRN,AMN,AFA)is far less than 2-4%, and your 2-4% improvement seems not a common result and is hard to obtain for pseudo labels generated with provided code.

AFA reports the CRF post-processing brings a gain of 2.2%. It looks AFA only used CRF for post-processing. For the other two methods you mentioned, I didn't find their reported results with and without post-processing. Only using CRF may bring a smaller gain like around 1%, it can also bring more like >= 3%, e.g., [[1]](https://openaccess.thecvf.com/content_CVPR_2020/papers/Araslanov_Single-Stage_Semantic_Segmentation_From_Image_Labels_CVPR_2020_paper.pdf) and [[2]](https://ojs.aaai.org/index.php/AAAI/article/view/6705)). For the benefits of using both multi-scale testing and CRF, you can refer to [this](https://ieeexplore.ieee.org/abstract/document/7913730).

**Comment by Gptao** [comment]
AFA's post-processing include MS too， it can be checked from [test_msc_flip](https://github.com/rulixiang/afa/blob/master/scripts/test_msc_flip.py),  It is true only using CRF may bring a smaller gain like around 1% from my tries,  as a beginner of WSSS, I still feel strange about the impact of post-processing on performance.  Thank you for your opinion.

**Comment by yijiangWang0316** [comment]
> [AFA的post-processing也包括MS，从test_msc_flip](https://github.com/rulixiang/afa/blob/master/scripts/test_msc_flip.py)可以查到，确实只用CRF可能带来的收益比较小，大概1%左右，作为WSSS的初学者，对于post-的影响还是很奇怪性能上的处理。谢谢您的意见。

您好，可以提供一下，python seg/train_seg.py 和 python seg/infer_seg.py训练时的命令吗？（一些路径问题，作者并没有给出详细的路径，我不知道如何匹配所需要的路径）做训练时，读取会一直出现错误，当img_path为JPEGImages时，提示在--seg_pgt_path：V2结果文件中，pgt-psa-rw文件里没有xxx的图片，对这里不是很明白。
对于您的热心回复，我将十分感谢您的帮助！

**Comment by yijiangWang0316** [comment]
> 

作者您好，可以提供一下，python seg/train_seg.py 和 python seg/infer_seg.py训练时的命令吗？（一些路径问题，作者并没有给出详细的路径，我不知道如何匹配所需要的路径）做训练时，读取会一直出现错误，当img_path为JPEGImages时，提示在--seg_pgt_path：V2结果文件中，pgt-psa-rw文件里没有xxx的图片，对这里不是很明白。
对于您的热心回复，我将十分感谢您的帮助！

**Comment by pipizhum** [comment]
> I am unable to reproduce the segmentation retraining results in your paper (71.9). I have used the pseudo-mask and pre-training weights you provided. I have also trained with exact reference to the default parameters you set in train_seg.py as well as run_seg.sh. But I only get 71.1 results. I am confused about this, can you tell me why? Thanks!
Can you provide instructions for all the training steps? I followed the author's instructions and the final segmentation result of mIoU was only 67.5%.

**Comment by pipizhum** [comment]
> I am unable to reproduce the segmentation retraining results in your paper (71.9). I have used the pseudo-mask and pre-training weights you provided. I have also trained with exact reference to the default parameters you set in train_seg.py as well as run_seg.sh. But I only get 71.1 results. I am confused about this, can you tell me why? Thanks!



> Can you provide instructions for all the training steps? I followed the author's instructions and the final segmentation result of mIoU was only 67.5%.

**Comment by pipizhum** [comment]
> I am unable to reproduce the segmentation retraining results in your paper (71.9). I have used the pseudo-mask and pre-training weights you provided. I have also trained with exact reference to the default parameters you set in train_seg.py as well as run_seg.sh. But I only get 71.1 results. I am confused about this, can you tell me why? Thanks!

> Can you provide instructions for all the training steps? I followed the author's instructions and the final segmentation result of mIoU was only 67.5%.

**Comment by yijiangWang0316** [comment]
> 
您好，可以提供关于run_seg.sh脚本中train_seg.py和infer_seg.py具体的命令行参数吗？或者提供一下你运行该脚本时的命令行参数，我不清楚具体应该填写什么才能达到论文中的结果。我并不能复现run_seg.sh脚本 。十分期待您的帮忙
祝您科研一切顺利~
@Yingyue-L
